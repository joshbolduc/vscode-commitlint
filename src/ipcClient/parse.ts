import { getContext } from '../getContext';
import type { ParseIpcRequest } from '../ipcTypes';
import { log } from '../log';
import { ensureWorkerManager } from './workerManager';

export const parse = async (
  text: string,
  path: string | undefined,
  parseOptions: {
    commentChar: string | undefined;
    verbose: boolean;
  },
) => {
  try {
    return await ensureWorkerManager().send<ParseIpcRequest>({
      type: 'parse',
      ...parseOptions,
      ...getContext(path),
      text,
    });
  } catch (e) {
    log(`Failed to parse message: ${String(e)}`);
  }
};
