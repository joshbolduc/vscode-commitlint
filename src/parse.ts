import { Commit } from '@commitlint/types';
import { importCommitlintParse } from './loadLibrary';

type KnownSection = 'header' | 'body' | 'footer' | 'scope' | 'subject' | 'type';

type Range = [start: number, end: number];

type SectionRanges = Partial<Record<KnownSection, Range>>;

function isCommentLine(line: string) {
  // TODO: this is actually configurable per git config. see:
  // https://stackoverflow.com/questions/22936252/escape-comment-character-in-git-commit-message
  const commentChar = '#';

  return line.startsWith(commentChar);
}

const LINE_BREAK = '\n';

function splitCommit(text: string) {
  return text.split(LINE_BREAK);
}

interface Offset {
  index: number;
  length: number;
}

function getCommitRanges(commit: Commit) {
  const text = commit.raw;

  const headerStart = text.indexOf(commit.header);
  const headerEnd = headerStart + commit.header.length;

  const ranges: SectionRanges = {
    header: [headerStart, headerEnd],
  };

  if (commit.body) {
    const bodyStart = text.indexOf(commit.body, headerEnd);
    const bodyEnd = bodyStart + commit.body.length;
    ranges.body = [bodyStart, bodyEnd];
  }

  if (commit.footer) {
    const footerStart = text.lastIndexOf(commit.footer);
    const footerEnd = footerStart + commit.footer.length;
    ranges.footer = [footerStart, footerEnd];
  }

  if (commit.type) {
    const typeStart = headerStart + commit.header.indexOf(commit.type);
    const typeEnd = headerStart + typeStart + commit.type.length;
    ranges.type = [typeStart, typeEnd];
  }

  if (commit.scope) {
    const scopeStart =
      headerStart + commit.header.indexOf(`(${commit.scope})`) + 1;
    const scopeEnd = headerStart + scopeStart + commit.scope.length;
    ranges.scope = [scopeStart, scopeEnd];
  }

  return ranges;
}

function isValidLine(line: string) {
  return !isCommentLine(line) && line !== '';
}

export async function parseCommit(text: string, path: string | undefined) {
  const lines = splitCommit(text);

  const firstContentLine = lines.findIndex(isValidLine);
  const lastContentLine = lines.reduceRight((prev, cur, i) => {
    if (prev === i && !isValidLine(cur)) {
      return i - 1;
    }

    return prev;
  }, lines.length - 1);

  const sanitizedText = lines
    .slice(firstContentLine, lastContentLine + 1)
    .filter((line) => !isCommentLine(line))
    .join(LINE_BREAK);

  const parse = importCommitlintParse(path);

  const commit = await parse(sanitizedText);
  const originalRanges = getCommitRanges(commit);

  const { offsets } = lines.reduce<{
    offsets: Offset[];
    index: number;
  }>(
    (acc, cur, i) => {
      const lineLength = cur.length + LINE_BREAK.length;

      const shouldIncludeLine =
        i >= firstContentLine && i <= lastContentLine && !isCommentLine(cur);

      if (shouldIncludeLine) {
        acc.index += lineLength;
      } else {
        acc.offsets.push({
          index: acc.index,
          length: lineLength,
        });
      }

      return acc;
    },
    { offsets: [], index: 0 },
  );

  const ranges = Object.entries(originalRanges).reduce<SectionRanges>(
    (acc, [section, sectionRange]) => {
      if (sectionRange) {
        const [start, end] = sectionRange;

        const adjustedRange = offsets.reduce<Range>(
          ([accStart, accEnd], { index, length }) => {
            const startOffset = index <= start ? length : 0;
            const endOffset = index < end ? length : 0;

            return [accStart + startOffset, accEnd + endOffset];
          },
          [start, end],
        );

        acc[section as keyof typeof ranges] = adjustedRange;
      }

      return acc;
    },
    {},
  );

  return { ranges, sanitizedText };
}
