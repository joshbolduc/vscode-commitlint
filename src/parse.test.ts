import { readFileSync } from 'fs';
import { join, resolve } from 'path';
import { fixturesPath, testLibRootPath } from '../test/util';
import { parseCommit } from './parse';

jest.mock('./log');
jest.mock('./settings');

describe('parse', () => {
  const versions = [11, 12, 13];

  const fixtures = [
    'commit-all-ranges.txt',
    'commit-leading-comment.txt',
    'commit-leading-comment-newlines.txt',
    'commit-trailing-comment.txt',
  ];

  versions.forEach((version) => {
    const libPath = resolve(testLibRootPath, `v${version}`);

    fixtures.forEach((fixture) => {
      it(`parses ${fixture} using v${version}`, async () => {
        const contents = readFileSync(join(fixturesPath, fixture)).toString();

        expect(await parseCommit(contents, libPath)).toMatchSnapshot();
      });
    });
  });
});
