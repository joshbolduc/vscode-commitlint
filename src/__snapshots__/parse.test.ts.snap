// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parse parses commit-all-ranges.txt using v11 1`] = `
Object {
  "ranges": Object {
    "body": Array [
      79,
      171,
    ],
    "footer": Array [
      173,
      184,
    ],
    "header": Array [
      25,
      52,
    ],
    "scope": Array [
      30,
      35,
    ],
    "type": Array [
      25,
      29,
    ],
  },
  "sanitizedText": "feat(scope): commit subject


Commit body text here.
Commit body text here.
Commit body text here.

Closes #123",
}
`;

exports[`parse parses commit-all-ranges.txt using v12 1`] = `
Object {
  "ranges": Object {
    "body": Array [
      79,
      171,
    ],
    "footer": Array [
      173,
      184,
    ],
    "header": Array [
      25,
      52,
    ],
    "scope": Array [
      30,
      35,
    ],
    "type": Array [
      25,
      29,
    ],
  },
  "sanitizedText": "feat(scope): commit subject


Commit body text here.
Commit body text here.
Commit body text here.

Closes #123",
}
`;

exports[`parse parses commit-all-ranges.txt using v13 1`] = `
Object {
  "ranges": Object {
    "body": Array [
      79,
      171,
    ],
    "footer": Array [
      173,
      184,
    ],
    "header": Array [
      25,
      52,
    ],
    "scope": Array [
      30,
      35,
    ],
    "type": Array [
      25,
      29,
    ],
  },
  "sanitizedText": "feat(scope): commit subject


Commit body text here.
Commit body text here.
Commit body text here.

Closes #123",
}
`;

exports[`parse parses commit-leading-comment.txt using v11 1`] = `
Object {
  "ranges": Object {
    "body": Array [
      18,
      27,
    ],
    "header": Array [
      10,
      16,
    ],
  },
  "sanitizedText": "Header

Body line",
}
`;

exports[`parse parses commit-leading-comment.txt using v12 1`] = `
Object {
  "ranges": Object {
    "body": Array [
      18,
      27,
    ],
    "header": Array [
      10,
      16,
    ],
  },
  "sanitizedText": "Header

Body line",
}
`;

exports[`parse parses commit-leading-comment.txt using v13 1`] = `
Object {
  "ranges": Object {
    "body": Array [
      18,
      27,
    ],
    "header": Array [
      10,
      16,
    ],
  },
  "sanitizedText": "Header

Body line",
}
`;

exports[`parse parses commit-leading-comment-newlines.txt using v11 1`] = `
Object {
  "ranges": Object {
    "body": Array [
      20,
      29,
    ],
    "header": Array [
      12,
      18,
    ],
  },
  "sanitizedText": "Header

Body line",
}
`;

exports[`parse parses commit-leading-comment-newlines.txt using v12 1`] = `
Object {
  "ranges": Object {
    "body": Array [
      20,
      29,
    ],
    "header": Array [
      12,
      18,
    ],
  },
  "sanitizedText": "Header

Body line",
}
`;

exports[`parse parses commit-leading-comment-newlines.txt using v13 1`] = `
Object {
  "ranges": Object {
    "body": Array [
      20,
      29,
    ],
    "header": Array [
      12,
      18,
    ],
  },
  "sanitizedText": "Header

Body line",
}
`;

exports[`parse parses commit-trailing-comment.txt using v11 1`] = `
Object {
  "ranges": Object {
    "body": Array [
      8,
      17,
    ],
    "header": Array [
      0,
      6,
    ],
  },
  "sanitizedText": "Header

Body line",
}
`;

exports[`parse parses commit-trailing-comment.txt using v12 1`] = `
Object {
  "ranges": Object {
    "body": Array [
      8,
      17,
    ],
    "header": Array [
      0,
      6,
    ],
  },
  "sanitizedText": "Header

Body line",
}
`;

exports[`parse parses commit-trailing-comment.txt using v13 1`] = `
Object {
  "ranges": Object {
    "body": Array [
      8,
      17,
    ],
    "header": Array [
      0,
      6,
    ],
  },
  "sanitizedText": "Header

Body line",
}
`;
